docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,113 +97,148 @@ The current multisig public keys use amino serialization to generate the address
 
-  // then concatenate them together
-  var joinedAddresses []byte
-  for addr := range addresses {
-    joinedAddresses := append(joinedAddresses, addr...)
-  }
+  // use the Composed function defined above
+  return address.Composed(prefix, keys)
+}
+```
 
-  // form the string prefix from the message name (cosmos.crypto.multisig.PubKey) and the threshold joined together
-  prefix := fmt.Sprintf("%s/%d", proto.MessageName(multisig), multisig.Threshold)
 
-  // use the standard AddressHash function
-  return AddressHash(prefix, joinedAddresses)
+### Account Types
+
+The Account Types used in various account classes SHOULD be unique for each class.
+Since both public keys and accounts are serialized in the state, we propose to use the protobuf message name string.
+
+Example: all public key types have a unique protobuf message type similar to:
+
+```proto
+package cosmos.crypto.sr25519;
+
+message PubKey {
+  bytes key = 1;
 }
-``` 
+```
+
+All protobuf messages have unique fully qualified names, in this example `cosmos.crypto.sr25519.PubKey`.
+These names are derived directly from .proto files in a standardized way and used
+in other places such as the type URL in `Any`s. We can easily obtain the name using
+`proto.MessageName(msg)`.
+
+
 
 ## Consequences
 
+### Backwards Compatibility
+
+This ADR is compatible to what was committed and directly supported in the SDK repository.
+
 ### Positive
-- a simple algorithm for generating addresses for new public keys and module accounts
+
+- a simple algorithm for generating addresses for new public keys, complex accounts and module accounts
+- the algorithm generalizes for _native composed keys_
+- increase security and collision resistance of addresses
+- the approach is extensible for future use-cases - one can use other address types, as long as they don't conflict with the address length specified here (20 or 32 bytes).
+- support multiple types accounts.
 
 ### Negative
+
 - addresses do not communicate key type, a prefixed approach would have done this
+- addresses are 60% longer and will consume more storage space
+- requires a refactor of KVStore store keys to handle variable length addresses
 
 ### Neutral
+
 - protobuf message names are used as key type prefixes
 
+
+## Further Discussions
+
+Some accounts can have a fixed name or may be constructed in other way (eg: module accounts). We were discussing an idea of an account with a predefined name (eg: `me.regen`), which could be used by institutions.
+Without going into details, this kind of addresses are compatible with the hash based addresses described here as long as they don't have the same length.
+More specifically, any special account address, must not have length equal to 20 byte nor 32 bytes.
+
+
 ## References
+
+* [Notes](https://hackmd.io/_NGWI4xZSbKzj1BkCqyZMw) from consulting meeting with [Alan Szepieniec](https://scholar.google.be/citations?user=4LyZn8oAAAAJ&hl=en).
```

Comment:
Could these notes be added as another document directly in the repository?  These seem like an important piece of context to preserve in association with this ADR and it isn't clear that hackmd will preserve this record with the same retention as the source repository.

Reply:
OK, so how about adding it here, as an _Appendix_ section?

Reply:
done